File: clients/go/internal/cli/gardencmd.go
package cli

import (
	"errors"
	"fmt"
	"math/big"
	"os"
	"path/filepath"
	"strings"
	"time"
	"unicode/utf8"

	"daisugi-verify/internal/garden"
	"daisugi-verify/internal/pathways"
	"daisugi-verify/internal/pmodel"
	"daisugi-verify/internal/pyjson"
	"daisugi-verify/internal/pystr"
)

var gardenerHelp = `Usage: daisugi gardener [OPTIONS] COMMAND [ARGS]...

  Lifecycle management for compiled pathways (prune, merge, status).

Commands:
  prune   Evict stale / failure-dominated pathways.
  merge   Collapse near-duplicate pathways.
  run     Run the full gardener pipeline (prune + merge).
  watch   Cron-friendly one-shot gardener.
  status  Report current store size, pathway activation stats, failure ratios.
`

func (e *Env) gardener(args []string) error {
	if len(args) == 0 || args[0] == "--help" {
		e.out("%s", gardenerHelp)
		return nil
	}
	sub, rest := args[0], args[1:]
	switch sub {
	case "prune":
		return e.gardenerPrune(rest)
	case "merge":
		return e.gardenerMerge(rest)
	case "run":
		return e.gardenerRun(rest)
	case "watch":
		return e.gardenerWatch(rest)
	case "status":
		return e.gardenerStatus(rest)
	}
	e.errf("Usage: daisugi gardener [OPTIONS] COMMAND [ARGS]...\nTry 'daisugi gardener --help' for help.\n\nError: No such command '%s'.\n", sub)
	return exit(2)
}

// nowSeconds is time.time().
func nowSeconds() float64 { return float64(time.Now().UnixNano()) / 1e9 }

// clickFloat reads a FLOAT option: Python's float() of the text.
func clickFloat(p *parsed, name string, def float64) (float64, error) {
	if !p.has(name) {
		return def, nil
	}
	raw := p.str(name, "")
	f, ok := pmodel.FloatFromString(raw)
	if !ok {
		return 0, &usageError{fmt.Sprintf("Invalid value for '%s': %s is not a valid float.", name, pystr.Repr(raw))}
	}
	return f, nil
}

// clickInt reads an INT option: Python's int() of the text.
func clickInt(p *parsed, name string, def int64) (int64, error) {
	if !p.has(name) {
		return def, nil
	}
	raw := p.str(name, "")
	n, ok := pyInt(raw)
	if !ok {
		return 0, &usageError{fmt.Sprintf("Invalid value for '%s': %s is not a valid int.", name, pystr.Repr(raw))}
	}
	return n, nil
}

func dryRunOpt() opt { return opt{names: []string{"--dry-run"}} }

func (e *Env) gardenStoreErr(cmd string, err error) error {
	var inv *pathways.Invalid
	if errors.As(err, &inv) {
		return e.fail(cmd, err)
	}
	return e.storeErr(cmd, err)
}

func pruneJSON(r garden.PruneReport, withDry bool, dry bool) *pyjson.Object {
	o := pyjson.NewObject().Set("removed_ids", strAny(r.RemovedIDs)).Set("kept_count", r.KeptCount)
	reasons := pyjson.NewObject()
	for _, id := range r.RemovedIDs {
		reasons.Set(id, r.Reasons[id])
	}
	o.Set("reasons", reasons)
	if withDry {
		o.Set("dry_run", dry)
	}
	return o
}

func pairsAny(pairs [][2]string) []any {
	out := make([]any, len(pairs))
	for i, p := range pairs {
		out[i] = []any{p[0], p[1]}
	}
	return out
}

func strAny(ss []string) []any {
	out := make([]any, len(ss))
	for i, s := range ss {
		out[i] = s
	}
	return out
}

func (e *Env) gardenerPrune(args []string) error {
	const cmd = "gardener prune"
	opts := []opt{dataDirOpt,
		{names: []string{"--max-idle-days"}, value: true, metavar: "FLOAT"},
		{names: []string{"--max-failure-ratio"}, value: true, metavar: "FLOAT"},
		{names: []string{"--min-activations"}, value: true, metavar: "INTEGER"},
		dryRunOpt(), {names: []string{"--json"}}}
	p, err := parseArgs(args, opts, 0)
	if err != nil {
		return e.usage(cmd, err)
	}
	if p.help {
		return e.cmdHelp(cmd, "", "Evict stale / failure-dominated pathways.", opts)
	}
	cfg := garden.DefaultPrune()
	if cfg.MaxIdleDays, err = clickFloat(p, "--max-idle-days", 30); err != nil {
		return e.usage(cmd, err)
	}
	if cfg.MaxFailureRatio, err = clickFloat(p, "--max-failure-ratio", 0.5); err != nil {
		return e.usage(cmd, err)
	}
	n, err := clickInt(p, "--min-activations", 5)
	if err != nil {
		return e.usage(cmd, err)
	}
	cfg.MinActivations = big.NewInt(n)
	s, err := e.openStore(cmd, p)
	if err != nil {
		return err
	}
	defer s.Close()
	dry := p.flag("--dry-run")
	rep, err := garden.Prune(s, cfg, dry, nowSeconds())
	if err != nil {
		return e.gardenStoreErr(cmd, err)
	}
	if p.flag("--json") {
		e.out("%s\n", pyjson.DumpsIndent(pruneJSON(rep, true, dry), 2, true))
		return nil
	}
	verb := "removed"
	if dry {
		verb = "would remove"
	}
	e.out("%s: %d (kept: %d)\n", verb, len(rep.RemovedIDs), rep.KeptCount)
	for _, id := range rep.RemovedIDs {
		e.out("  %s — %s\n", id, rep.Reasons[id])
	}
	return nil
}

func (e *Env) gardenerMerge(args []string) error {
	const cmd = "gardener merge"
	opts := []opt{dataDirOpt, {names: []string{"--similarity"}, value: true, metavar: "FLOAT"},
		dryRunOpt(), {names: []string{"--json"}}}
	p, err := parseArgs(args, opts, 0)
	if err != nil {
		return e.usage(cmd, err)
	}
	if p.help {
		return e.cmdHelp(cmd, "", "Collapse near-duplicate pathways.", opts)
	}
	cfg := garden.DefaultMerge()
	if cfg.SimilarityThreshold, err = clickFloat(p, "--similarity", 0.92); err != nil {
		return e.usage(cmd, err)
	}
	s, err := e.openStore(cmd, p)
	if err != nil {
		return err
	}
	defer s.Close()
	dry := p.flag("--dry-run")
	rep, err := garden.Merge(s, cfg, dry)
	if err != nil {
		return e.gardenStoreErr(cmd, err)
	}
	if p.flag("--json") {
		o := pyjson.NewObject().Set("merged_pairs", pairsAny(rep.MergedPairs)).
			Set("kept_ids", strAny(rep.KeptIDs)).Set("removed_ids", strAny(rep.RemovedIDs)).Set("dry_run", dry)
		e.out("%s\n", pyjson.DumpsIndent(o, 2, true))
		return nil
	}
	verb := "merged"
	if dry {
		verb = "would merge"
	}
	e.out("%s: %d pair(s)\n", verb, len(rep.MergedPairs))
	for _, pr := range rep.MergedPairs {
		e.out("  %s  <-  %s\n", pr[0], pr[1])
	}
	return nil
}

// runGardener is gardener.run_gardener: prune, then merge.
func runGardener(s *pathways.Store, dry bool) (garden.PruneReport, garden.MergeReport, error) {
	pr, err := garden.Prune(s, garden.DefaultPrune(), dry, nowSeconds())
	if err != nil {
		return pr, garden.MergeReport{}, err
	}
	mr, err := garden.Merge(s, garden.DefaultMerge(), dry)
	return pr, mr, err
}

func (e *Env) gardenerRun(args []string) error {
	const cmd = "gardener run"
	opts := []opt{dataDirOpt, dryRunOpt(), {names: []string{"--json"}}}
	p, err := parseArgs(args, opts, 0)
	if err != nil {
		return e.usage(cmd, err)
	}
	if p.help {
		return e.cmdHelp(cmd, "", "Run the full gardener pipeline (prune + merge).", opts)
	}
	s, err := e.openStore(cmd, p)
	if err != nil {
		return err
	}
	defer s.Close()
	dry := p.flag("--dry-run")
	pr, mr, err := runGardener(s, dry)
	if err != nil {
		return e.gardenStoreErr(cmd, err)
	}
	if p.flag("--json") {
		o := pyjson.NewObject().
			Set("prune", pruneJSON(pr, false, false)).
			Set("merge", pyjson.NewObject().Set("merged_pairs", pairsAny(mr.MergedPairs)).Set("kept_ids", strAny(mr.KeptIDs))).
			Set("dry_run", dry)
		e.out("%s\n", pyjson.DumpsIndent(o, 2, true))
		return nil
	}
	verb := ""
	if dry {
		verb = "would"
	}
	e.out("prune: %s removed %d, kept %d\n", verb, len(pr.RemovedIDs), pr.KeptCount)
	e.out("merge: %s merged %d pair(s)\n", verb, len(mr.MergedPairs))
	return nil
}

// readStamp is the last-run stamp as the oracle reads it: a float, or 0
// when the file is absent or its text is not one.
func readStamp(path string) (float64, error) {
	raw, err := os.ReadFile(path)
	if errors.Is(err, os.ErrNotExist) {
		return 0, nil
	}
	if err != nil {
		return 0, err
	}
	if !utf8.Valid(raw) {
		return 0, errors.New("UnicodeDecodeError: the stamp file is not UTF-8")
	}
	text := strings.ReplaceAll(strings.ReplaceAll(string(raw), "\r\n", "\n"), "\r", "\n")
	f, ok := pmodel.FloatFromString(pystr.Strip(text))
	if !ok {
		return 0, nil
	}
	return f, nil
}

func (e *Env) gardenerWatch(args []string) error {
	const cmd = "gardener watch"
	opts := []opt{dataDirOpt,
		{names: []string{"--min-interval"}, value: true, metavar: "INTEGER", help: "Skip if the last run is newer than this many seconds."},
		{names: []string{"--force"}, help: "Ignore the min-interval check."}, dryRunOpt()}
	p, err := parseArgs(args, opts, 0)
	if err != nil {
		return e.usage(cmd, err)
	}
	if p.help {
		return e.cmdHelp(cmd, "", "Cron-friendly one-shot gardener. Skips if last run is within --min-interval.", opts)
	}
	minInterval, err := clickInt(p, "--min-interval", 3600)
	if err != nil {
		return e.usage(cmd, err)
	}
	dir := p.str("--data-dir", filepath.Join(e.home, ".opendaisugi"))
	stamp := filepath.Join(dir, ".gardener-last-run")
	now := nowSeconds()
	last, err := readStamp(stamp)
	if err != nil {
		return e.failPy(cmd, err)
	}
	elapsed := now - last
	if !p.flag("--force") && elapsed < float64(minInterval) {
		o := pyjson.NewObject().Set("skipped", true).Set("reason", "min_interval_not_elapsed").
			Set("elapsed_s", pyjson.Round(elapsed, 1)).Set("min_interval_s", pyjson.Int{Text: fmt.Sprint(minInterval)})
		e.out("%s\n", pyjson.Dumps(o, true))
		return nil
	}
	s, err := e.openStore(cmd, p)
	if err != nil {
		return err
	}
	defer s.Close()
	dry := p.flag("--dry-run")
	pr, mr, err := runGardener(s, dry)
	if err != nil {
		return e.gardenStoreErr(cmd, err)
	}
	if !dry {
		if err := os.MkdirAll(dir, 0o777); err != nil {
			return e.failPy(cmd, err)
		}
		if err := os.WriteFile(stamp, []byte(garden.FormatF(now, 3)), 0o666); err != nil {
			return e.failPy(cmd, err)
		}
	}
	o := pyjson.NewObject().Set("skipped", false).Set("ran_at", now).Set("dry_run", dry).
		Set("prune", pyjson.NewObject().Set("removed", len(pr.RemovedIDs)).Set("kept", pr.KeptCount)).
		Set("merge", pyjson.NewObject().Set("merged", len(mr.MergedPairs)))
	e.out("%s\n", pyjson.Dumps(o, true))
	return nil
}

func (e *Env) gardenerStatus(args []string) error {
	const cmd = "gardener status"
	opts := []opt{dataDirOpt, {names: []string{"--json"}}}
	p, err := parseArgs(args, opts, 0)
	if err != nil {
		return e.usage(cmd, err)
	}
	if p.help {
		return e.cmdHelp(cmd, "", "Report current store size, pathway activation stats, failure ratios.", opts)
	}
	s, err := e.openStore(cmd, p)
	if err != nil {
		return err
	}
	defer s.Close()
	all, err := s.ReadAll(func(string) bool { return false })
	if err != nil {
		return e.gardenStoreErr(cmd, err)
	}
	if p.flag("--json") {
		items := []any{}
		for _, pw := range all {
			o := pw.Obj
			items = append(items, pyjson.NewObject().Set("id", o.Value("id")).
				Set("hit_count", o.Value("hit_count")).Set("failure_count", o.Value("failure_count")).
				Set("last_activation_at", o.Value("last_activation_at")))
		}
		o := pyjson.NewObject().Set("count", len(all)).Set("pathways", items)
		e.out("%s\n", pyjson.DumpsIndent(o, 2, true))
		return nil
	}
	e.out("count: %d\n", len(all))
	for _, pw := range all {
		o := pw.Obj
		h, _ := new(big.Int).SetString(o.Value("hit_count").(pyjson.Int).Text, 10)
		f, _ := new(big.Int).SetString(o.Value("failure_count").(pyjson.Int).Text, 10)
		total := new(big.Int).Add(h, f)
		ratio := 0.0
		if total.Sign() != 0 {
			ratio, _ = new(big.Rat).SetFrac(f, total).Float64()
		}
		e.out("  %s  hits=%s  fails=%s  fail_ratio=%s\n", pw.ID(), h, f, garden.FormatF(ratio, 2))
	}
	return nil
}

File: clients/rust/src/cli/gardencmd.rs
//! `daisugi gardener prune|merge|run|watch|status`: the pathway store's
//! life cycle, with the flags, output, exit codes and files of the Python
//! CLI.

use super::pathwayscmd::{py_int, DATA_DIR};
use super::{exit, parse_args, Env, Opt, Parsed, Res, Stop};
use crate::garden::{self, format_f, int_of, MergeConfig, MergeReport, PruneConfig, PruneReport};
use crate::gate::argparse::py_float;
use crate::gate::py::text::repr;
use crate::gate::pyjson::{dumps, dumps_indent, round, Object, Value};
use crate::pathways::store::Store;
use crate::pathways::PwErr;

const GARDENER_HELP: &str = "Usage: daisugi gardener [OPTIONS] COMMAND [ARGS]...

  Lifecycle management for compiled pathways (prune, merge, status).

Commands:
  prune   Evict stale / failure-dominated pathways.
  merge   Collapse near-duplicate pathways.
  run     Run the full gardener pipeline (prune + merge).
  watch   Cron-friendly one-shot gardener.
  status  Report current store size, pathway activation stats, failure ratios.
";

const DRY_RUN: Opt = Opt::flag(&["--dry-run"], "Report what would change; write nothing.");
const JSON: Opt = Opt::flag(&["--json"], "Machine-readable JSON output.");

/// `time.time()`.
pub fn now_seconds() -> f64 {
    match std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH) {
        Ok(d) => d.as_secs_f64(),
        Err(_) => 0.0,
    }
}

fn strs(ss: &[String]) -> Value {
    Value::List(ss.iter().map(|s| Value::Str(s.clone())).collect())
}

fn pairs(ps: &[(String, String)]) -> Value {
    Value::List(ps.iter().map(|(a, b)| Value::List(vec![Value::Str(a.clone()), Value::Str(b.clone())])).collect())
}

fn prune_json(r: &PruneReport, dry: Option<bool>) -> Object {
    let ids: Vec<String> = r.removed.iter().map(|(id, _)| id.clone()).collect();
    let mut o = Object::new();
    o.set("removed_ids", strs(&ids));
    o.set("kept_count", r.kept as i64);
    let mut reasons = Object::new();
    for (id, why) in &r.removed {
        reasons.set(id, why.as_str());
    }
    o.set("reasons", reasons);
    if let Some(d) = dry {
        o.set("dry_run", d);
    }
    o
}

impl Env {
    pub(super) fn gardener(&mut self, args: &[String]) -> Res {
        if args.is_empty() || args[0] == "--help" {
            self.out(GARDENER_HELP);
            return Ok(());
        }
        let (sub, rest) = (args[0].as_str(), &args[1..]);
        match sub {
            "prune" => self.gardener_prune(rest),
            "merge" => self.gardener_merge(rest),
            "run" => self.gardener_run(rest),
            "watch" => self.gardener_watch(rest),
            "status" => self.gardener_status(rest),
            _ => {
                self.errf(&format!(
                    "Usage: daisugi gardener [OPTIONS] COMMAND [ARGS]...\nTry 'daisugi gardener --help' for help.\n\n\
                     Error: No such command '{sub}'.\n"
                ));
                exit(2)
            }
        }
    }

    /// A FLOAT option: Python's `float()` of the text.
    pub(super) fn click_float(&mut self, cmd: &str, p: &Parsed, name: &str, def: f64) -> Result<f64, Stop> {
        if !p.has(name) {
            return Ok(def);
        }
        let raw = p.str(name, "");
        match py_float(&raw) {
            Some(f) => Ok(f),
            None => Err(self.usage_stop(cmd, format!("Invalid value for '{name}': {} is not a valid float.", repr(&raw)))),
        }
    }

    /// An INTEGER option: Python's `int()` of the text.
    pub(super) fn click_int(&mut self, cmd: &str, p: &Parsed, name: &str, def: i128) -> Result<i128, Stop> {
        if !p.has(name) {
            return Ok(def);
        }
        let raw = p.str(name, "");
        match py_int(&raw) {
            Some(n) => Ok(n),
            None => Err(self.usage_stop(cmd, format!("Invalid value for '{name}': {} is not a valid int.", repr(&raw)))),
        }
    }

    fn gardener_prune(&mut self, args: &[String]) -> Res {
        const CMD: &str = "gardener prune";
        let opts = [
            DATA_DIR,
            Opt::val(&["--max-idle-days"], "FLOAT", "Evict a pathway idle longer than this."),
            Opt::val(&["--max-failure-ratio"], "FLOAT", "Evict a pathway failing more often than this."),
            Opt::val(&["--min-activations"], "INTEGER", "Activations before the failure ratio counts."),
            DRY_RUN,
            JSON,
        ];
        let p = parse_args(args, &opts, 0).map_err(|m| self.usage_stop(CMD, m))?;
        if p.help {
            return self.cmd_help(CMD, "", "Evict stale / failure-dominated pathways.", &opts);
        }
        let cfg = PruneConfig {
            max_idle_days: self.click_float(CMD, &p, "--max-idle-days", 30.0)?,
            max_failure_ratio: self.click_float(CMD, &p, "--max-failure-ratio", 0.5)?,
            min_activations: self.click_int(CMD, &p, "--min-activations", 5)?,
        };
        let s = self.open_store(CMD, &p)?;
        let dry = p.flag("--dry-run");
        let rep = garden::prune(&s, &cfg, dry, now_seconds()).map_err(|e| self.pw_err(CMD, e))?;
        if p.flag("--json") {
            self.out(&format!("{}\n", dumps_indent(&Value::Obj(prune_json(&rep, Some(dry))), 2, true)));
            return Ok(());
        }
        let verb = if dry { "would remove" } else { "removed" };
        let mut b = format!("{verb}: {} (kept: {})\n", rep.removed.len(), rep.kept);
        for (id, why) in &rep.removed {
            b.push_str(&format!("  {id} \u{2014} {why}\n"));
        }
        self.out(&b);
        Ok(())
    }

    fn gardener_merge(&mut self, args: &[String]) -> Res {
        const CMD: &str = "gardener merge";
        let opts = [
            DATA_DIR,
            Opt::val(&["--similarity"], "FLOAT", "Cosine similarity at which two pathways merge."),
            DRY_RUN,
            JSON,
        ];
        let p = parse_args(args, &opts, 0).map_err(|m| self.usage_stop(CMD, m))?;
        if p.help {
            return self.cmd_help(CMD, "", "Collapse near-duplicate pathways.", &opts);
        }
        let cfg = MergeConfig { similarity_threshold: self.click_float(CMD, &p, "--similarity", 0.92)?, ..Default::default() };
        let s = self.open_store(CMD, &p)?;
        let dry = p.flag("--dry-run");
        let rep = garden::merge(&s, &cfg, dry).map_err(|e| self.pw_err(CMD, e))?;
        if p.flag("--json") {
            let mut o = Object::new();
            o.set("merged_pairs", pairs(&rep.merged_pairs));
            o.set("kept_ids", strs(&rep.kept_ids));
            o.set("removed_ids", strs(&rep.removed_ids));
            o.set("dry_run", dry);
            self.out(&format!("{}\n", dumps_indent(&Value::Obj(o), 2, true)));
            return Ok(());
        }
        let verb = if dry { "would merge" } else { "merged" };
        let mut b = format!("{verb}: {} pair(s)\n", rep.merged_pairs.len());
        for (w, l) in &rep.merged_pairs {
            b.push_str(&format!("  {w}  <-  {l}\n"));
        }
        self.out(&b);
        Ok(())
    }

    /// `gardener.run_gardener`: prune, then merge, each with its defaults.
    fn run_gardener(&mut self, cmd: &str, s: &Store, dry: bool) -> Result<(PruneReport, MergeReport), Stop> {
        let pr = garden::prune(s, &PruneConfig::default(), dry, now_seconds()).map_err(|e| self.pw_err(cmd, e))?;
        let mr = garden::merge(s, &MergeConfig::default(), dry).map_err(|e| self.pw_err(cmd, e))?;
        Ok((pr, mr))
    }

    fn gardener_run(&mut self, args: &[String]) -> Res {
        const CMD: &str = "gardener run";
        let opts = [DATA_DIR, DRY_RUN, JSON];
        let p = parse_args(args, &opts, 0).map_err(|m| self.usage_stop(CMD, m))?;
        if p.help {
            return self.cmd_help(CMD, "", "Run the full gardener pipeline (prune + merge).", &opts);
        }
        let s = self.open_store(CMD, &p)?;
        let dry = p.flag("--dry-run");
        let (pr, mr) = self.run_gardener(CMD, &s, dry)?;
        if p.flag("--json") {
            let mut m = Object::new();
            m.set("merged_pairs", pairs(&mr.merged_pairs));
            m.set("kept_ids", strs(&mr.kept_ids));
            let mut o = Object::new();
            o.set("prune", prune_json(&pr, None));
            o.set("merge", m);
            o.set("dry_run", dry);
            self.out(&format!("{}\n", dumps_indent(&Value::Obj(o), 2, true)));
            return Ok(());
        }
        let verb = if dry { "would" } else { "" };
        self.out(&format!(
            "prune: {verb} removed {}, kept {}\nmerge: {verb} merged {} pair(s)\n",
            pr.removed.len(),
            pr.kept,
            mr.merged_pairs.len()
        ));
        Ok(())
    }

    fn gardener_watch(&mut self, args: &[String]) -> Res {
        const CMD: &str = "gardener watch";
        let opts = [
            DATA_DIR,
            Opt::val(&["--min-interval"], "INTEGER", "Skip if the last run is newer than this many seconds."),
            Opt::flag(&["--force"], "Ignore the min-interval check."),
            DRY_RUN,
        ];
        let p = parse_args(args, &opts, 0).map_err(|m| self.usage_stop(CMD, m))?;
        if p.help {
            return self.cmd_help(CMD, "", "Cron-friendly one-shot gardener. Skips if last run is within --min-interval.", &opts);
        }
        let min_interval = self.click_int(CMD, &p, "--min-interval", 3600)?;
        let dir = p.str("--data-dir", &format!("{}/.opendaisugi", self.home));
        let stamp = format!("{dir}/.gardener-last-run");
        let now = now_seconds();
        let last = read_stamp(&stamp).map_err(|e| self.pw_err(CMD, e))?;
        let elapsed = now - last;
        if !p.flag("--force") && elapsed < min_interval as f64 {
            let mut o = Object::new();
            o.set("skipped", true);
            o.set("reason", "min_interval_not_elapsed");
            o.set("elapsed_s", round(elapsed, 1));
            o.set("min_interval_s", Value::Int(min_interval.to_string()));
            self.out(&format!("{}\n", dumps(&Value::Obj(o), true)));
            return Ok(());
        }
        let s = self.open_store(CMD, &p)?;
        let dry = p.flag("--dry-run");
        let (pr, mr) = self.run_gardener(CMD, &s, dry)?;
        if !dry {
            let write = std::fs::create_dir_all(&dir).and_then(|_| std::fs::write(&stamp, format_f(now, 3)));
            if let Err(e) = write {
                return Err(self.pw_err(CMD, PwErr::Io(e)));
            }
        }
        let mut pj = Object::new();
        pj.set("removed", pr.removed.len() as i64);
        pj.set("kept", pr.kept as i64);
        let mut mj = Object::new();
        mj.set("merged", mr.merged_pairs.len() as i64);
        let mut o = Object::new();
        o.set("skipped", false);
        o.set("ran_at", now);
        o.set("dry_run", dry);
        o.set("prune", pj);
        o.set("merge", mj);
        self.out(&format!("{}\n", dumps(&Value::Obj(o), true)));
        Ok(())
    }

    fn gardener_status(&mut self, args: &[String]) -> Res {
        const CMD: &str = "gardener status";
        let opts = [DATA_DIR, JSON];
        let p = parse_args(args, &opts, 0).map_err(|m| self.usage_stop(CMD, m))?;
        if p.help {
            return self.cmd_help(CMD, "", "Report current store size, pathway activation stats, failure ratios.", &opts);
        }
        let s = self.open_store(CMD, &p)?;
        let all = s.read_all(&|_| false).map_err(|e| self.pw_err(CMD, e))?;
        if p.flag("--json") {
            let items: Vec<Value> = all
                .iter()
                .map(|pw| {
                    let mut x = Object::new();
                    for k in ["id", "hit_count", "failure_count", "last_activation_at"] {
                        x.set(k, pw.obj.value(k).clone());
                    }
                    Value::Obj(x)
                })
                .collect();
            let mut o = Object::new();
            o.set("count", all.len() as i64);
            o.set("pathways", Value::List(items));
            self.out(&format!("{}\n", dumps_indent(&Value::Obj(o), 2, true)));
            return Ok(());
        }
        let mut b = format!("count: {}\n", all.len());
        for pw in &all {
            let (h, f) = (int_of(pw.obj.value("hit_count")), int_of(pw.obj.value("failure_count")));
            let total = h + f;
            let ratio = if total == 0 { 0.0 } else { garden::true_div(f, total) };
            b.push_str(&format!("  {}  hits={h}  fails={f}  fail_ratio={}\n", pw.id(), format_f(ratio, 2)));
        }
        self.out(&b);
        Ok(())
    }
}

/// The last-run stamp as the oracle reads it: a float, or 0 when the file
/// is absent or its text is not one.
pub fn read_stamp(path: &str) -> Result<f64, PwErr> {
    let raw = match std::fs::read(path) {
        Ok(b) => b,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(0.0),
        Err(e) => return Err(PwErr::Io(e)),
    };
    let text = match String::from_utf8(raw) {
        Ok(t) => t,
        Err(_) => return Err(PwErr::Invalid("UnicodeDecodeError: the stamp file is not UTF-8".into())),
    };
    let text = text.replace("\r\n", "\n").replace('\r', "\n");
    Ok(py_float(crate::gate::py::text::strip(&text)).unwrap_or(0.0))
}
